File: content/examples/other/leetcode/problems/top-k-frequent-elements/example-2/solution.js
function topKFrequentBucket(nums, k) {
  const freq = new Map();
  for (const x of nums) freq.set(x, (freq.get(x) || 0) + 1);

  const buckets = Array.from({ length: nums.length + 1 }, () => []);
  for (const [num, count] of freq.entries()) buckets[count].push(num);

  const out = [];
  for (let c = buckets.length - 1; c >= 0 && out.length < k; c -= 1) {
    for (const num of buckets[c]) {
      out.push(num);
      if (out.length === k) break;
    }
  }
  return out;
}

if (require.main === module) {
  console.log(topKFrequentBucket([1, 1, 1, 2, 2, 3], 2));
}

module.exports = { topKFrequentBucket };

File: content/examples/other/leetcode/problems/top-k-frequent-elements/example-2/solution.py
def top_k_frequent_bucket(nums, k):
    freq = {}
    for x in nums:
        freq[x] = freq.get(x, 0) + 1

    buckets = [[] for _ in range(len(nums) + 1)]
    for num, count in freq.items():
        buckets[count].append(num)

    out = []
    for c in range(len(buckets) - 1, -1, -1):
        for num in buckets[c]:
            out.append(num)
            if len(out) == k:
                return out
    return out


if __name__ == "__main__":
    print(top_k_frequent_bucket([1, 1, 1, 2, 2, 3], 2))
